doc/adr/2022-11-16-option-precedence-hierarchy.md

```diff
@@ -0,0 +1,28 @@
+**Title**: Establish how multiple values for the same option are reconciled.
+
+**Status**: proposed
+
+**Context**: There are a variety of ways that options can be customized for a
+given rpc call. Both `Client` and `Connection` classes can be created with a set
+of `google::cloud::Options`. Individual function calls in a `Client` can be
+passed `google::cloud::Options`. Some environment variables can be set that
+such that the variable value sets an Option associated with that environment
+variable. Lastly, GCS `Request` types inherit from various Option templates
+which allow the corresponding option to be set in an instance of the `Request`.
+
+When the same Option is set, with different values, via more than one of these
+mechanisms, there needs to be a uniform process of determining which value to
+use.
+
+**Decision**: The following hierarchy will be followed when determining which
+value to use when the same Option is set via multiple mechanisms:
+
+1. Environment variables
```

Comment:
Thinking ahead... taking `UserProjectOption` as an example.... I think the changes are just:
* Stop defaulting the option
* Instead store an `absl::optional<std::string> user_project_env_` in the class that uses the option
* Check that member variable, before looking in the current options.
* Update unit tests

That's not so bad.